examples/seismic/tti/operators.py

```diff
@@ -17,6 +17,137 @@ def ForwardOperator(model, source, receiver, time_order=2, space_order=4,
+        Hzr = (first_derivative(Gzr * ang1 * ang2,
+                                dim=x, side=centered, order=order1,
+                                matvec=transpose) +
+               first_derivative(Gzr * ang1 * ang3,
+                                dim=y, side=centered, order=order1,
+                                matvec=transpose) +
+               first_derivative(Gzr * ang0,
+                                dim=z, side=centered, order=order1,
+                                matvec=transpose))
+
+    else:
+        # Gx1p = -(ang1 * u.dx + ang0 * u.dy)
+        # Gz1r = -(ang0 * v.dx - ang1 * v.dy)
+        # Call first-derivative instead to get the wanted order
+        Gx = -(ang1 * first_derivative(u, dim=x, side=centered, order=order1) +
+               ang0 * first_derivative(u, dim=y, side=centered, order=order1))
+        Gxx1 = (first_derivative(Gx * ang1, dim=x,
+                                 side=centered, order=order1,
+                                 matvec=transpose) +
+                first_derivative(Gx * ang0, dim=y,
+                                 side=centered, order=order1,
+                                 matvec=transpose))
+        Hp = u.laplace - Gxx1
+        # Hzr
+        Gx = -(ang1 * first_derivative(v, dim=x, side=centered, order=order1) +
+               ang0 * first_derivative(v, dim=y, side=centered, order=order1))
+        Hzr = (first_derivative(Gx * ang1, dim=x,
+                                side=centered, order=order1,
+                                matvec=transpose) +
+               first_derivative(Gx * ang0, dim=y,
+                                side=centered, order=order1,
+                                matvec=transpose))
+
+    s = t.spacing
+    # Stencils
+    stencilp = 1.0 / (2.0 * m + s * damp) * \
+        (4.0 * m * u + (s * damp - 2.0 * m) *
+         u.backward + 2.0 * s**2 * (epsilon * Hp + delta * Hzr))
+    stencilr = 1.0 / (2.0 * m + s * damp) * \
+        (4.0 * m * v + (s * damp - 2.0 * m) *
+         v.backward + 2.0 * s**2 * (delta * Hp + Hzr))
+    first_stencil = Eq(u.forward, stencilp)
+    second_stencil = Eq(v.forward, stencilr)
+    stencils = [first_stencil, second_stencil]
+
+    # SOurce and receivers
+    stencils += src.inject(field=u.forward, expr=src * dt * dt / m,
+                           offset=model.nbpml)
+    stencils += src.inject(field=v.forward, expr=src * dt * dt / m,
+                           offset=model.nbpml)
+    stencils += rec.interpolate(expr=u + v, offset=model.nbpml)
+    # Add substitutions for spacing (temporal and spatial)
+    subs = dict([(t.spacing, dt)] + [(time.spacing, dt)] +
+                [(i.spacing, model.get_spacing()[j]) for i, j
+                 in zip(u.indices[1:], range(len(model.shape)))])
+    # Operator
+    return Operator(stencils, subs=subs, name='ForwardTTI', **kwargs)
+
+
+def ForwardOperator_legacy(model, source, receiver, time_order=2, space_order=4,
```

Comment:
So here's a rule of thumb: any time you do a copy-paste which contains more than one line, that there's a non-negligible chance that things can be done in a much better way. 

Do I get it right : the difference between ``ForwardOperator`` and ``ForwardOperator_legacy`` is only in the symbolics, while the other lines are exactly the same. If that's the case, you have to refrain from copy-pasting and think about how we can factorize some code. In this case, one possibility is to have one function ``ForwardOperator`` that does the setup, and then two other functions -- one function per operator -- that take the same parameters (the symbols ``u, m, ...`` and return the symbolic equations. In fact, you could apply this same reasoning to the 2D specifications. 

Then you could have ``ForwardOperator(...)`` taking an additional name, representing what type of operator one wants to execute (give them significant names, not "legacy"); you can also have a dictionary mapping names to functions. Then you can imagine the rest; if the provided name is not in the dictionary, you raise an error; otherwise, you fetch the corresponding function